File: PlantHereNodeJs/Presentation/PlantHere.WebAPI/Controllers/ProductController.js
//  SERVICE
const ProductService = require('../../../Core/PlantHere.Aplication/Services/ProductService').ProductService


const GetProductsByPageQuery = require('../../../Core/PlantHere.Aplication/RequestResponseModels/Product/Queries/GetProductsByPage/GetProductsByPageQuery').GetProductsByPageQuery
const GetProductsByCategoryIdAndPageQuery = require('../../../Core/PlantHere.Aplication/RequestResponseModels/Product/Queries/GetProductByCategoryIdAndPage/GetProductByCategoryIdAndPageQuery').GetProductsByCategoryIdAndPageQuery
const GetProductByIdQuery = require('../../../Core/PlantHere.Aplication/RequestResponseModels/Product/Queries/GetProductById/GetProductByIdQuery').GetProductByIdQuery
const DeleteProductCommand = require('../../../Core/PlantHere.Aplication/RequestResponseModels/Product/Commands/DeleteProduct/DeleteProductCommand').DeleteProductCommand
const UpdateProductCommand = require('../../../Core/PlantHere.Aplication/RequestResponseModels/Product/Commands/UpdateProduct/UpdateProductCommand').UpdateProductCommand
const CreateProductCommand = require('../../../Core/PlantHere.Aplication/RequestResponseModels/Product/Commands/CreateProduct/CreateProductCommand').CreateProductCommand
const GetProductsCountQuery = require('../../../Core/PlantHere.Aplication/RequestResponseModels/Product/Queries/GetProductsCount/GetProductsCountQuery').GetProductsCountQuery

const service = new ProductService();

//Results
const { CustomResult } = require('../Results/CustomResult')


const getProductsCount = async (req, res) => {
    res.json(CustomResult.Success(await service.getProductsCount(new GetProductsCountQuery())))
}

const getProductsByPage = async (req, res) => {
    res.json(CustomResult.Success(await service.getProductsByPage(new GetProductsByPageQuery(req.params.page, req.params.pageSize))))
}

const getProductsByCategoryIdAndPage = async (req, res) => {
    res.json(CustomResult.Success(await service.getProductsByCategoryIdAndPage(new GetProductsByCategoryIdAndPageQuery(req.body.page, req.body.pageSize), req.params.categoryId)))
}

const getProductById = async (req, res) => {
    res.json(CustomResult.Success(await service.getProductsById(new GetProductByIdQuery(req.params.id))))
}

const deleteProduct = async (req, res) => {
    res.json(CustomResult.Success(await service.deleteProduct(new DeleteProductCommand(req.params.id))))
}

const createProduct = async (req, res) => {
    res.json(CustomResult.Success(await service.createProduct(new CreateProductCommand(req.body.name, req.body.description, req.body.price, req.body.sellerId, req.body.categoryId, req.body.discount, req.body.stock))))
}

const updateProduct = async (req, res) => {
    res.json(CustomResult.Success(await service.updateProduct(new UpdateProductCommand(req.body.id, req.body.name, req.body.description, req.body.price, req.body.sellerId, req.body.categoryId, req.body.discount))))
}

module.exports = {
    getProductsCount,
    getProductsByPage,
    getProductsByCategoryIdAndPage,
    getProductById,
    deleteProduct,
    createProduct,
    updateProduct
}

File: PlantHere/PlantHere/Controllers/ProductController.cs
using MediatR;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using PlantHere.Application.CQRS.Product.Commands.CreateProduct;
using PlantHere.Application.CQRS.Product.Commands.CreateProductsIndexES;
using PlantHere.Application.CQRS.Product.Commands.DeleteProduct;
using PlantHere.Application.CQRS.Product.Commands.UpdateProduct;
using PlantHere.Application.CQRS.Product.Queries.GetProductByUniqueId;
using PlantHere.Application.CQRS.Product.Queries.GetProductsByPage;
using PlantHere.Application.CQRS.Product.Queries.GetProductsCount;
using PlantHere.Application.CQRS.Product.Queries.GetProductsES;
using PlantHere.WebAPI.CustomResults;
using System.Net;

namespace PlantHere.WebAPI.Controllers
{
    [Route("products")]
    public class ProductController : CustomBaseController
    {
        private readonly IMediator _mediator;

        public ProductController(IMediator mediator)
        {
            _mediator = mediator;
        }

        /// <summary>
        /// Get Product By Unique Id
        /// </summary>
        /// <param name="uniqueId"></param>
        [AllowAnonymous]
        [HttpGet("{uniqueId}")]
        public async Task<CustomResult<GetProductByUniqueIdQueryResult>> GetProductByUniqueId(string uniqueId)
        {
            return CustomResult<GetProductByUniqueIdQueryResult>.Success((int)HttpStatusCode.OK, await _mediator.Send(new GetProductByUniqueIdQuery(uniqueId)));
        }

        /// <summary>
        /// Get Products By Page
        /// </summary>
        /// <param name="page"></param>
        /// <param name="pageSize"></param>
        [AllowAnonymous]
        [HttpGet("{page}/{pageSize}")]
        public async Task<CustomResult<IEnumerable<GetProductsByPageQueryResult>>> GetProductsByPage(int page, int pageSize)
        {
            return CustomResult<IEnumerable<GetProductsByPageQueryResult>>.Success((int)HttpStatusCode.OK, await _mediator.Send(new GetProductsByPageQuery(page, pageSize)));
        }

        /// <summary>
        /// Get Products Count
        /// </summary>
        [AllowAnonymous]
        [HttpGet("count")]
        public async Task<CustomResult<GetProductsCountQueryResult>> GetProductsCount()
        {
            return CustomResult<GetProductsCountQueryResult>.Success((int)HttpStatusCode.OK, await _mediator.Send(new GetProductsCountQuery()));
        }

        /// <summary>
        /// Update Product
        /// </summary>
        /// <param name="command"></param>
        [Authorize(Roles = "superadmin,seller")]
        [HttpPut]
        public async Task<CustomResult<GetProductByUniqueIdQueryResult>> UpdateProduct(UpdateProductCommand command)
        {
            await _mediator.Send(command);
            return CustomResult<GetProductByUniqueIdQueryResult>.Success((int)HttpStatusCode.NoContent);
        }

        /// <summary>
        /// Create Product
        /// </summary>
        /// <param name="command"></param>
        [Authorize(Roles = "superadmin,seller")]
        [HttpPost]
        public async Task<CustomResult<CreateProductCommandResult>> CreateProduct(CreateProductCommand command)
        {
            return CustomResult<CreateProductCommandResult>.Success((int)HttpStatusCode.OK, await _mediator.Send(command));
        }

        /// <summary>
        /// Get Products ES
        /// </summary>
        /// <param name="keyword"></param>
        [AllowAnonymous]
        [HttpGet("elastic-search/{keyword}")]
        public async Task<CustomResult<List<GetProductsESQueryResult>>> GetProductsES(string keyword)
        {
            var products = await _mediator.Send(new GetProductsESQuery(keyword));
            return CustomResult<List<GetProductsESQueryResult>>.Success((int)HttpStatusCode.OK, products);
        }

        /// <summary>
        /// Create Products Index ES
        /// </summary>
        [AllowAnonymous]
        [HttpGet("elastic-search-index")]
        public async Task<CustomResult<CreateProductsIndexESCommandResult>> CreateProductsIndexES()
        {
            return CustomResult<CreateProductsIndexESCommandResult>.Success((int)HttpStatusCode.OK, await _mediator.Send(new CreateProductsIndexESCommand()));
        }

        /// <summary>
        /// Delete Product
        /// </summary>
        /// <param name="id"></param>
        [Authorize(Roles = "superadmin")]
        [HttpDelete("{id}")]
        public async Task<CustomResult<DeleteProductCommand>> DeleteProduct(int id)
        {
            await _mediator.Send(new DeleteProductCommand(id));
            return CustomResult<DeleteProductCommand>.Success((int)HttpStatusCode.NoContent);
        }
    }
}
